numpy/core/src/multiarray/mapping.c

```diff
@@ -252,7 +252,26 @@ unpack_indices(PyObject *index, PyObject **result, npy_intp result_n)
 }
 
 /**
- * Prepare an npy_index_object from the python slicing object.
+ * If arg is a callable, apply it to self.
+ *
+ * @returns arg, arg(self), or NULL. NULL is returned if arg(self) fails
+ *      (if the callable takes more than a single argument, for example).
+ */
+static NPY_INLINE PyObject *
+maybe_apply_callable(PyArrayObject *self, PyObject *arg)
+{
+    if (PyFunction_Check(arg)) {
+        // NOTE: Starting with Python 3.9, we can instead do
+        //   PyObject *result = (PyObject *) PyObject_CallOneArg(arg, self);
+        PyObject *self_as_tuple = PyTuple_Pack(1, self);
+        PyObject *result = (PyObject *) PyObject_CallObject(arg, self_as_tuple);
+        return (result == NULL) ? NULL : result;
+    }
+    return arg;
```

Comment:
```suggestion
    Py_INCREF(arg);
    return arg;
```
as the `if` branch returns a new reference, so the else needs to as well.

This also means you need to decrement the refcount in the calling function.

Reply:
> as the if branch returns a new reference, so the else needs to as well.

Thanks for this! 

> This also means you need to decrement the refcount in the calling function.

I added `Py_DECREF(index)` in the three exit points of `prepare_index`

Reply:
*four exit points